server/src/main/java/io/crate/execution/engine/indexing/ShardingUpsertExecutor.java

```diff
@@ -105,11 +102,11 @@ public class ShardingUpsertExecutor
     private volatile boolean createPartitionsRequestOngoing = false;
     private final Predicate<UpsertResults> earlyTerminationCondition;
     private final Function<UpsertResults, Throwable> earlyTerminationExceptionGenerator;
-    private final Map<String, Consumer<IndexItem>> validatorsCache = new HashMap<>();
+    private final Runnable validatorsCacheCleaner;
 
     ShardingUpsertExecutor(ClusterService clusterService,
-                           DocTableInfo tableInfo,
-                           @Nullable List<Reference> allTargetColumns,
+                           @Nullable BiConsumer<String, IndexItem> constraintsChecker,
+                           @Nullable Runnable validatorsCacheCleaner,
```

Comment:
We could just pass in dummy a runnable `() -> {}` and consumer, so any `NULL` check can be avoided,